gstreamer/src/buffer.rs

```diff
@@ -579,6 +579,79 @@ lazy_static! {
         *BUFFER_COPY_METADATA | ::BufferCopyFlags::MEMORY;
 }
 
+#[cfg(feature = "ser_de")]
+pub(crate) mod serde {
+    use serde::de::{Deserialize, Deserializer};
+    use serde::ser::{Serialize, Serializer, SerializeStruct};
+    use serde_bytes::{Bytes, ByteBuf};
+
+    use Buffer;
+    use BufferFlags;
+    use BufferRef;
+    use ClockTime;
+
+    impl<'a> Serialize for BufferRef {
+        fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+            let mut datetime = serializer.serialize_struct("Buffer", 6)?;
+            datetime.serialize_field("pts", &self.get_pts())?;
+            datetime.serialize_field("dts", &self.get_dts())?;
+            datetime.serialize_field("duration", &self.get_duration())?;
+            datetime.serialize_field("offset", &self.get_offset())?;
+            datetime.serialize_field("offset_end", &self.get_offset_end())?;
+            datetime.serialize_field("flags", &self.get_flags())?;
+            {
+                let data = self.map_readable().unwrap();
+                datetime.serialize_field("buffer", &Bytes::new(data.as_slice()))?;
+            }
+            datetime.end()
+        }
+    }
+
+    impl<'a> Serialize for Buffer {
+        fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+            self.as_ref().serialize(serializer)
+        }
+    }
+
+    #[derive(Deserialize)]
+    struct BufferDe {
+        pts: ClockTime,
+        dts: ClockTime,
+        duration: ClockTime,
+        offset: u64,
+        offset_end: u64,
+        flags: BufferFlags,
+        buffer: ByteBuf,
+    }
+
+    impl From<BufferDe> for Buffer {
+        fn from(mut buf_de: BufferDe) -> Self {
+            // Move the `ByteBuff`'s byte array into the `gst::Buffer`
+            // Are we really avoiding copies with that?
+            let drained_buffer: &mut Vec<u8> = buf_de.buffer.as_mut();
+            let mut buffer =
+                Buffer::from_slice(drained_buffer.drain(..).collect::<Vec<u8>>()).unwrap();
```

Comment:
drain and collect will do a copy, so you're not really gaining anything here

Reply:
All right, I'll see what I can do.